### PR TITLE
sec: post-merge cleanup — rate-limiting + URL prefix + SSRF inline + coverage +4.6pp + backlog

### PR DESCRIPTION
**Pull request by ConcordDev**
## Why

Followup to PR #330 (the "first all-green CI in 56 days" merge). This PR closes the remaining issues surfaced by the wave 6 CodeQL review + does a re-audit pass on the over-excluded c8 test files.

### 1. Real CodeQL fixes
- **`server/routes/auth.js`** — `authRateLimitMiddleware` was applied to only `/register` and `/login`. Now wraps every credential-mutation route: `/refresh`, `/logout`, `/revoke-all-sessions`, `/change-password`, `/api-keys/:keyId/rotate`. Closes `js/missing-rate-limiting`.
- **`server/lib/chat-agent.js`** — `browse_url` + `browser_act` tools used `String(url).startsWith("http")` which matches `"httpevil://attacker"`. Replaced with URL parser + `protocol === "http:" || "https:"` exact check. Closes `js/incomplete-url-substring-sanitization`.
- **`server/lib/mcp-client.js`** — Inline re-validation of the URL immediately before fetch makes the SSRF sanitizer dataflow visible to CodeQL's taint tracker (it doesn't follow `validateSafeFetchUrl()` across function boundaries). Runtime behavior unchanged — defense-in-depth re-check.

### 2. Test-infra cleanup
- **`.github/workflows/ci.yml`** — Re-included 3 test files I'd over-excluded from c8 in PR #330 (`adversarial-critical-endpoints`, `edge-cases-critical-paths`, `error-paths`). Local repro confirms they pass cleanly under c8 when CI's `--test-timeout` is unset. Net: **+4.6pp statement coverage, +0.55pp function coverage**.

### 3. Backlog documentation
- **`docs/security/test-infra-backlog.md`** — 5 remaining items with explicit "done when" criteria:
  1. E2E suite (800+ specs red ~2mo, `continue-on-error: true`) — 1-2 weeks
  2. Remaining 2 c8 exclusions (auth-security + storage-parity)
  3. detectors-suite env-sensitive flake
  4. Function coverage thin margin (6.42pp above floor after this PR)
  5. CodeQL plateau at ~45-49 (mostly documented FPs)

Each item has an owner, estimated effort, recommended approach. Re-review cadence anchored to the existing `codeql-exclusions-review.yml` 90-day cron.

## Coverage delta

| Metric | Before PR | After this PR | Floor |
|---|---|---|---|
| Statements | 58.18% | **62.79%** | 45% |
| Branches | 72.27% | **70.6%** | 45% |
| Functions | 38.87% | **39.42%** | 33% |
| Lines | 58.18% | **62.79%** | 45% |

## Test plan

- [x] `cd server && npm run lint:ci` — 0 errors / 0 warnings
- [x] `cd server && npx c8 --check-coverage ... node --test` (full CI command) — exit 0, all 4 thresholds met
- [x] `auth-security` + `agent-marathon-mcp` touched-area tests — 39/39 pass
- [ ] CI gates green

## Risk

Low. Each change is surgical:
- Rate-limit middleware is a wrap-in — failed rate-limit returns 429 (legitimate clients unaffected)
- URL parser is stricter — invalid URLs now rejected before any network I/O (was the intent already)
- SSRF re-validation is redundant defense — same validator already ran in connect()
- c8 re-inclusion increased coverage, didn't fail

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
_Generated by [Claude Code](https://claude.ai/code/session_0143LVhbrKR9563RtvxzjVL5)_

### CODE REVIEW

File: server/lib/mcp-client.js [hunk 1]
```diff
@@ -338,6 +338,16 @@ export class MCPClient {
...
     try {
+      // Re-validate the URL inline immediately before fetch — makes the
+      // SSRF sanitizer dataflow visible to CodeQL's taint tracker, which
+      // doesn't follow validateSafeFetchUrl() across function boundaries.
+      // The connect() handshake already validated via validateSafeFetchUrl
+      // and stashed _urlSafety; this is a defense-in-depth re-check that
+      // ALSO satisfies CodeQL js/server-side-request-forgery.
+      const reChecked = await validateSafeFetchUrl(this._url);
+      if (!reChecked.ok) {
+        throw new Error(`SSRF re-validation failed: ${reChecked.error}`);
```
[the remark this answers is not in the document]
Comment (ConcordDev) [reply server/lib/mcp-client.js]: Fixed in 3c8a1c58. Split into two paths: pinned-IP fetch (happy path) skips per-request DNS lookup entirely so transient resolver outages can't break in-flight RPCs; the no-`_urlSafety` fallback still re-validates inline so we never fire blind. SSRF posture preserved because the pinned IP was validated at connect-time and `fetchWithPinnedIp` pins it explicitly.

---
_Generated by [Claude Code](https://claude.ai/code)_
